givenergy_modbus/client/commands.py

```diff
@@ -332,9 +330,16 @@ def set_active_power_rate(target: int) -> list[TransparentRequest]:
     """Set max inverter power rate"""
     return [WriteHoldingRegisterRequest(RegisterMap.ACTIVE_POWER_RATE, target)]
 
-def set_calibrate_battery_soc() -> list[TransparentRequest]:
-    """Set the inverter to recalibrate the battery state of charge estimation."""
-    return [WriteHoldingRegisterRequest(RegisterMap.SOC_FORCE_ADJUST, 1)]
+def set_calibrate_battery_soc(val: int) -> list[TransparentRequest]:
+    """Set the inverter to recalibrate the battery state of charge estimation.
+    0- Stop
+    1- Start
+    3- Charge only
+    """
+    if val in (0,1,3):
```

Comment:
Hmm... okay, current check assumes a range of valid values, whereas this wants a discrete set of valid values.

One bodge to extend that would be if 'valid' is a 2-tuple, it's interpreted as a a range, as now, otherwise it's a list of values.  If you want to allow exactly two discrete values, that can be bodged by repeating one.  eg  (1,5)  allows 1 <= x <= 5   whereas (1,5,5)  allows x in (1,5,5)  (ie  x ==1 or x ==5)

A single allowed value, such as for reboot, could then be done as  (100,)   but it's just as easy to do it as (the interval)  (100, 100)

But if this is really to trigger a battery calibration cycle, is it something we're allowed to offer ?  I though only GivEnergy were allowed to trigger calibrations ?

Reply:
Calibration is automatic from the inverter firmware, so although there's 7 states (from memory) the only actual options are stop (0) start (1) & charge only (3) 

Charge Only is a new one, this may be the elusive slow calibration sometimes run) 
The rest are just indication of current stage sonitll gradually rattle through each state before setting the battery capacity. 

As for whether we expose it or not we'll probably need to check with GE, however we do need read visibility of it as things like Predbat monitor for it at relinquish control for the duration as it would screw it up.

Reply:
Reboot MUST have soem form of protection around it. Anything other than valid value (100?) completely wipes the inverter leaving you with a brick. Bad times all round.

Reply:
In discussion with GE, It’s ok to allow battery calibration via GivTCP, only those three states

Reply:
> Calibration ...
> As for whether we expose it or not we'll probably need to check with GE, however we do need read visibility of it as things like Predbat monitor for it and relinquish control for the duration as it would screw it up.

Yes, I was really asking about whether we wanted to be offering access to set it. Read access should be less of an issue.

Reply:
> Reboot MUST have some form of protection around it. Anything other than valid value (100?) completely wipes the inverter leaving you with a brick. Bad times all round.

Even though I'd want to avoid duplication of any regiser definitions, I do wonder if we should make an exception for this one, and have the constructor for WriteHoldingRegisterRequest() absolutely forbid creating a request to write anything other than 100 to that register.

Reply:
> Hmm... okay, current check assumes a range of valid values, whereas this wants a discrete set of valid values.
> 
> One bodge to extend that would be if 'valid' is a 2-tuple, it's interpreted as a a range, as now, otherwise it's a list of values. If you want to allow exactly two discrete values, that can be bodged by repeating one. eg (1,5) allows 1 <= x <= 5 whereas (1,5,5) allows x in (1,5,5) (ie x ==1 or x ==5)
> 

A more general approach (with less risk of cockups) would be for 'valid' to be either a 2-tuple (min, max)  or a callable. Then it can be a lambda with whatever test is required.  so eg this could be

       "soc_force_adjust": Def(C.uint16, BatteryCalibrationStage, HR(29), valid = lambda val : val in (0,1,3 )),

then the lookup can use  callable(valid)  to determine whether 'valid' is a lambda or a simple (min, max) range.

Come to think of it, that could remove the need for the special-casing I currently have which assumes that if the upper limit is 2359 then it must be a time, and therefore needs an additional check that minutes < 60.  That can be done more explicitly using a validity funtion instead.